File: tools/tf_ws/src/tf_python/tf_python/static_tf_broadcaster.py
import rclpy
from rclpy.node import Node
from tf2_ros import StaticTransformBroadcaster # 导入静态变换广播器
from geometry_msgs.msg import TransformStamped # 导入TransformStamped消息类型
from tf_transformations import quaternion_from_euler # 欧拉角转四元数
import math

class StaticTFBroadcasterNode(Node):
    def __init__(self):
        super().__init__("static_tf_broadcaster")
        self.tfbrosdcaster = StaticTransformBroadcaster(self)
        # 发布静态TF点
        self.publish_static_tf()

    
    def publish_static_tf(self):
        """
        手眼转换
        发布静态TF点
        从 base_link 到 camera_link
        """
        tf = TransformStamped()
        tf.header.frame_id = "base_link"
        tf.child_frame_id = "camera_link"
        tf.header.stamp = self.get_clock().now().to_msg()

        tf.transform.translation.x = 0.5
        tf.transform.translation.y = 0.3
        tf.transform.translation.z = 0.6

        q = quaternion_from_euler(math.radians(180),0,0)
        tf.transform.rotation.x = q[0]
        tf.transform.rotation.y = q[1]
        tf.transform.rotation.z = q[2]
        tf.transform.rotation.w = q[3]

        self.tfbrosdcaster.sendTransform(tf)
        self.get_logger().info("发布静态TF点: 从 base_link 到 camera_link，{}".format(tf))

def main(args=None):
    rclpy.init(args=args)
    node = StaticTFBroadcasterNode()
    rclpy.spin(node) 
    rclpy.shutdown()

File: tools/tf_ws/src/tf_cpp/src/static_tf_broadcaster.cpp
#include "rclcpp/rclcpp.hpp"
#include "tf2_ros/static_transform_broadcaster.h"
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "tf2/LinearMath/Quaternion.h"  // 用于四元数计算
#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp" // 用于转换消息类型
#include "tf2_ros/static_transform_broadcaster.hpp"

class StaticTFBroadcaster : public rclcpp::Node
{
private:
    std::shared_ptr<tf2_ros::StaticTransformBroadcaster> static_broadcaster_;
public:
    StaticTFBroadcaster():Node("static_tf_broadcaster"){
        static_broadcaster_ = std::make_shared<tf2_ros::StaticTransformBroadcaster>(this);
        this->publish_tf();
    }

    void publish_tf(){
        // 定义一个静态变换
        geometry_msgs::msg::TransformStamped static_transform_stamped;

        static_transform_stamped.header.stamp = this->get_clock()->now();
        static_transform_stamped.header.frame_id = "map";
        static_transform_stamped.child_frame_id = "target_point";

        // 设置平移部分
        static_transform_stamped.transform.translation.x = 5.0;
        static_transform_stamped.transform.translation.y = 3.0;
        static_transform_stamped.transform.translation.z = 0.0;

        // 设置旋转部分（使用四元数表示）
        tf2::Quaternion quat;
        quat.setRPY(0, 0, 60*M_PI/180.0); // Roll, Pitch, Yaw
        static_transform_stamped.transform.rotation.x = quat.x();
        static_transform_stamped.transform.rotation.y = quat.y();
        static_transform_stamped.transform.rotation.z = quat.z();
        static_transform_stamped.transform.rotation.w = quat.w();

        // 发布静态变换
        static_broadcaster_->sendTransform(static_transform_stamped);

        RCLCPP_INFO(this->get_logger(), "Static Transform from 'map' to 'target_point' published.");
    }
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<StaticTFBroadcaster>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}



